File: tests/test_preferences.py
"""User preferences: pure dataclass round-trip + sanitization."""
from __future__ import annotations

from ads1292_studio.ui_qt.preferences import Preferences


def test_round_trips_through_dict() -> None:
    prefs = Preferences(
        port="/dev/cu.usbmodem", mode="Raw Record",
        save_csv=True, save_h5=False, save_xlsx=True,
        window="12 s", gain="2x", speed="50 mm/s",
    )
    assert Preferences.from_dict(prefs.to_dict()) == prefs


def test_from_dict_ignores_unknown_and_fills_defaults() -> None:
    prefs = Preferences.from_dict({"port": "/dev/x", "bogus": 1})
    assert prefs.port == "/dev/x"
    # defaults preserved for missing keys
    assert prefs.save_csv is True
    assert prefs.save_h5 is True
    assert prefs.save_xlsx is False
    assert prefs.window == "8 s"


def test_from_dict_coerces_types() -> None:
    prefs = Preferences.from_dict({"save_csv": 0, "save_xlsx": 1, "gain": 1})
    assert prefs.save_csv is False
    assert prefs.save_xlsx is True
    assert prefs.gain == "1"


def test_empty_dict_is_all_defaults() -> None:
    assert Preferences.from_dict({}) == Preferences()

File: tests/cpp/test_preferences.cpp
// tests/cpp/test_preferences.cpp
// Unit tests for ads1292::gui::Preferences struct + to_map/from_map.
// No Qt dependency — pure struct + std::map.

#include "catch.hpp"
#include "ads1292/gui/Preferences.h"

using ads1292::gui::Preferences;

TEST_CASE("Preferences defaults", "[preferences]") {
    Preferences p;
    REQUIRE(p.port     == "");
    REQUIRE(p.mode     == "Live Monitor");
    REQUIRE(p.save_csv == true);
    REQUIRE(p.save_h5  == true);
    REQUIRE_FALSE(p.save_xlsx);
    REQUIRE(p.window   == "8 s");
    REQUIRE(p.gain     == "1x");
    REQUIRE(p.speed    == "25 mm/s");
}

TEST_CASE("Preferences round-trip via to_map/from_map", "[preferences]") {
    Preferences a;
    a.port      = "/dev/cu.x";
    a.save_xlsx = true;
    a.window    = "4 s";

    auto m = a.to_map();
    auto b = Preferences::from_map(m);

    REQUIRE(b.port      == "/dev/cu.x");
    REQUIRE(b.save_xlsx == true);
    REQUIRE(b.window    == "4 s");
    REQUIRE(b.save_h5   == true);   // default true survives round-trip
    REQUIRE(b.save_csv  == true);   // default true survives round-trip
    REQUIRE(b.mode      == "Live Monitor");
}

TEST_CASE("Preferences from_map with missing keys keeps defaults", "[preferences]") {
    auto c = Preferences::from_map({{"port", "/dev/y"}});
    REQUIRE(c.port   == "/dev/y");
    REQUIRE(c.mode   == "Live Monitor");
    REQUIRE(c.save_h5 == true);
    REQUIRE_FALSE(c.save_xlsx);
    REQUIRE(c.window == "8 s");
}

TEST_CASE("Preferences from_map bool parsing", "[preferences]") {
    // "true" → true
    REQUIRE(Preferences::from_map({{"save_xlsx", "true"}}).save_xlsx == true);
    // "false" → false
    REQUIRE_FALSE(Preferences::from_map({{"save_h5", "false"}}).save_h5);
    // unknown string → false
    REQUIRE_FALSE(Preferences::from_map({{"save_xlsx", "nonsense"}}).save_xlsx);
    // truthy aliases
    REQUIRE(Preferences::from_map({{"save_xlsx", "1"}}).save_xlsx);
    REQUIRE(Preferences::from_map({{"save_xlsx", "yes"}}).save_xlsx);
    REQUIRE(Preferences::from_map({{"save_xlsx", "on"}}).save_xlsx);
    // case-insensitive
    REQUIRE(Preferences::from_map({{"save_xlsx", "True"}}).save_xlsx);
    REQUIRE(Preferences::from_map({{"save_xlsx", "YES"}}).save_xlsx);
}

TEST_CASE("Preferences to_map has 8 entries and encodes bools as strings", "[preferences]") {
    Preferences a;
    a.port      = "/dev/cu.x";
    a.save_xlsx = true;
    a.window    = "4 s";

    auto m = a.to_map();
    REQUIRE(m.size()              == 8);
    REQUIRE(m.at("save_xlsx")     == "true");
    REQUIRE(m.at("save_h5")      == "true");
    REQUIRE(m.at("save_csv")     == "true");
    REQUIRE(m.at("port")         == "/dev/cu.x");
    REQUIRE(m.at("mode")         == "Live Monitor");
    REQUIRE(m.at("window")       == "4 s");
    REQUIRE(m.at("gain")         == "1x");
    REQUIRE(m.at("speed")        == "25 mm/s");
}
